### PR TITLE
Support simple C++ linters

### PR DESCRIPTION
**Pull request by vonschultz**
Support basic C++ linters/formatters, such as clang-format or cpplint.
Tools that require a compile-commands.json are not yet in scope.

Clang-format will insist that the configuration file is called
".clang-format" and that it resides in a parent directory, and will
not tolerate any mention of the file on the command line. It wants the
`"-i"` flag to format code in place. To accomodate this, permit
specifying both config_str and config, and don't mention the config
file on the command line unless config_opt is specified. In the
clang-format case, you would run with `config_str="-i"` and
`config="//.clang-format"` (to bring the file into the sandbox).

### CODE REVIEW

File: generator.bzl [hunk 1]
```diff
@@ -1,3 +1,5 @@
+# Changed 2020 by Zenseact AB
```
Comment (vonschultz) [inline generator.bzl:1]: This is to comply with the Apache 2.0 license §4b, "You must cause any modified files to carry prominent notices stating that You changed the files."

If the file had the standard boilerplate header, http://www.apache.org/licenses/LICENSE-2.0#apply, I think the normal way to comply with §4b would be to just add another line with a copyright statement, but I don't feel I can just insert a Zenseact copyright statement if there wasn't any to begin with. Do you want this type of notice, and how would you like it to look?

File: rules.bzl [hunk 2]
```diff
@@ -13,7 +15,7 @@ def _linter_impl(ctx):
     return [
         LinterInfo(
             executable_path=ctx.attr.executable_path,
-            executable=ctx.executable.executable,
+            executable=ctx.attr.executable,
```
Comment (thundergolfer) [inline rules.bzl:18]: I suppose either way it's the same `File` object? What's the reason for the change?

**Review by thundergolfer — APPROVED** [review]
Thanks for the addition

File: rules.bzl (hunk 2 above)
Comment (vonschultz) [reply rules.bzl:18]: No, we're changing it from a [`File`](https://docs.bazel.build/versions/master/skylark/lib/File.html) object (`ctx.executable.executable`) to a [`Target`](https://docs.bazel.build/versions/master/skylark/lib/Target.html) object. You will find that `ctx.attr.executable[DefaultInfo].files_to_run.executable == ctx.executable.executable`.

We're using this with `clang-format` and `cpplint`. With `clang-format` you have a binary that needs at least two shared objects to be available in the sandbox. We have `cpplint` in a `py_binary`, and then it wanted an entire runfiles structure. In either case, we need more than just a file. We also need the runtime dependencies of the executable. We solve this with `ctx.resolve_tool(tools = [linter[LinterInfo].executable])` (see [ctx.resolve_tools](https://docs.bazel.build/versions/master/skylark/lib/ctx.html#resolve_tools)), which requires a sequence of  [`Target`](https://docs.bazel.build/versions/master/skylark/lib/Target.html) objects.
